File: sdk/php/src/Signer/HmacSigner.php
<?php

namespace HJTPX\Captcha\Signer;

class HmacSigner
{
    protected $secretKey;

    public function __construct(string $secretKey)
    {
        $this->secretKey = $secretKey;
    }

    public function sign(string $data): string
    {
        return hash_hmac('sha256', $data, $this->secretKey);
    }

    public function verify(string $data, string $signature): bool
    {
        $expectedSignature = $this->sign($data);
        return hash_equals($expectedSignature, $signature);
    }
}

File: sdk/csharp/src/Hjtpx.Captcha.Sdk/Signer/HmacSigner.cs
using System.Security.Cryptography;
using System.Text;

namespace Hjtpx.Captcha.Signer;

public class HmacSigner
{
    private readonly byte[] _secretKeyBytes;

    public HmacSigner(string secretKey)
    {
        _secretKeyBytes = Encoding.UTF8.GetBytes(secretKey);
    }

    public string Sign(string data)
    {
        using var hmac = new HMACSHA256(_secretKeyBytes);
        byte[] hashBytes = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
        return Convert.ToBase64String(hashBytes);
    }

    public bool Verify(string data, string signature)
    {
        string expectedSignature = Sign(data);
        return expectedSignature == signature;
    }
}
